hkdf/src/hkdf.rs

```diff
@@ -27,19 +27,26 @@ impl<D> Hkdf<D>
           D::OutputSize: ArrayLength<u8>
 {
     #[deprecated(since="0.4.0", note="renamed to extract, with arguments in reversed order")]
-    pub fn new(ikm: &[u8], salt: &[u8]) -> Hkdf<D> {
+    pub fn new(ikm: &[u8], salt: Option<&[u8]>) -> Hkdf<D> {
         Self::extract(salt, ikm)
     }
 
     /// The RFC5869 HKDF-Extract operation
-    pub fn extract(salt: &[u8], ikm: &[u8]) -> Hkdf<D> {
+    pub fn extract(salt: Option<&[u8]>, ikm: &[u8]) -> Hkdf<D> {
         // The hmac-0.5 MAC trait (which provides new()) is now defined to
         // return a Result, apparently to support things like CMAC which
         // require a specific key length. As far as I can tell, HMAC in
         // particular can only return an Ok(), since HMAC accepts any length
         // of bytes as its key. So we use unwrap() here, rather than exposing
         // the error to our caller. This might change in a future version.
-        let mut hmac = Hmac::<D>::new(salt).unwrap();
+        let s = if !salt.is_some() {
+            //Option::from(D::OutputSize as &[u8])
```

Comment:
@warner @newpavlov Any ideas for the syntax here? 

I guess what we want here is something like python:

```py
hash_len = hash().digest_size
if salt == None or len(salt) == 0:
    salt = bytearray((0,) * hash_len)
```

Reply:
Here I think here we can write something like this:
```Rust
let mut hmac = match salt {
    Some(s) => Hmac::<D>::new(salt),
    None => Hmac::<D>::new(&GenericArray<u8, D::OutputSize>::default()),
}.expect("HMAC can accept keys of any size");
```

Reply:
I updated to 0.11 of generay_array and getting

```rs
error[E0423]: expected value, found struct `generic_array::GenericArray`
  --> hkdf/src/hkdf.rs:44:37
   |
44 |             None => Hmac::<D>::new(&generic_array::GenericArray<u8, D::OutputSize>::default()),
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ did you mean `generic_array::GenericArray { /* fields */ }`?

error[E0423]: expected value, found builtin type `u8`
  --> hkdf/src/hkdf.rs:44:65
   |
44 |             None => Hmac::<D>::new(&generic_array::GenericArray<u8, D::OutputSize>::default()),
   |                                                                 ^^ not a value


```

with 

```rs
        let mut hmac = match salt {
            Some(s) => Hmac::<D>::new(salt.unwrap()),
            None => Hmac::<D>::new(&generic_array::GenericArray<u8, D::OutputSize>::default()),
        }.expect("HMAC can accept keys of any size").unwrap();
```

Reply:
Below we do have a: 
```
let mut prev: Option<generic_array::GenericArray<u8, <D as digest::FixedOutput>::OutputSize>> = None;
```

Going to try that

Reply:
Also you can depend on `digest` and `generic-array` implicitly:
```Rust
use hmac::digest::Digest;
use hmac::digest::generic_array::GenericArray;
```
This way you can remove those crates from your `Cargo.toml` and versions alwayswill be in sync.